File: app/src/main/java/com/example/practice/view/fragment/ProjectFragment.java
package com.example.practice.view.fragment;

import android.os.Bundle;

import com.example.practice.adapter.ProjectPageAdapter;
import com.example.practice.base.BaseFragment;
import com.example.practice.bean.ProjectListBean;
import com.example.practice.bean.ProjectPageBean;
import com.example.practice.config.Constants;
import com.example.practice.databinding.FragmentProjectBinding;
import com.example.practice.viewmodel.MainViewModel;
import com.example.practice.widge.MySlidingTabLayout;
import com.wljy.mvvmlibrary.annotation.Event;

import java.util.ArrayList;
import java.util.List;

import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;

/**
 * A simple {@link Fragment} subclass.
 * create an instance of this fragment.
 */
public class ProjectFragment extends BaseFragment<FragmentProjectBinding>{

    MySlidingTabLayout tableLayout;
    ViewPager viewPager;

    private MainViewModel mainViewModel;
    private List<ProjectListBean> projectListBeans;
    private List<ProjectPageBean> projectPageBeans;

    @Override
    public void initViewModel(){
        mainViewModel = registerViewModel(MainViewModel.class);
    }

    @Override
    public void initView(Bundle bundle){
        super.initView(bundle);
        tableLayout = binding.tableLayout;
        viewPager = binding.viewPager;
    }

    @Override
    public void initListener(){
        super.initListener();

    }

    @Override
    public void getRemoteData(){
        mainViewModel.getProject();
    }

    @Event(key = {Constants.GET_PROJECT_LIST, Constants.REQUEST_ERROR})
    public void event(String key, Object value){
        if(key.equals(Constants.GET_PROJECT_LIST)){
            projectListBeans = (List<ProjectListBean>) value;
            changePageFragment();
            relateVpAndTab();
        }else if(key.equals(Constants.REQUEST_ERROR)){
        }
    }

    private void relateVpAndTab(){
        //      https://www.jianshu.com/p/8fdd6f2719b0
        ProjectPageAdapter pageAdapter = new ProjectPageAdapter(getChildFragmentManager(), projectPageBeans);
        String titles[] = new String[projectPageBeans.size()];
        for(int i = 0; i < projectPageBeans.size(); i++){
            titles[i] = projectListBeans.get(i).getName();
        }
        viewPager.setAdapter(pageAdapter);
        tableLayout.setViewPager(viewPager, titles);
    }

    private List<ProjectPageBean> changePageFragment(){
        projectPageBeans = new ArrayList<>();
        for(ProjectListBean projectListBean : projectListBeans){
            ProjectPageBean projectPageBean = new ProjectPageBean();
            projectPageBean.setId(projectListBean.getId());
            projectPageBean.setTitle(projectListBean.getName());
            projectPageBean.setFragment(ProjectPageFragment.newInstance(projectListBean.getId()));
            projectPageBeans.add(projectPageBean);
        }
        return projectPageBeans;
    }
}

File: app/src/main/java/com/example/practice/view/fragment/ProjectFragment.kt
package com.example.practice.view.fragment

import android.os.Bundle
import androidx.viewpager.widget.ViewPager
import com.example.practice.adapter.ProjectPageAdapter
import com.example.practice.base.BaseFragment
import com.example.practice.bean.ProjectListBean
import com.example.practice.bean.ProjectPageBean
import com.example.practice.config.Constants
import com.example.practice.databinding.FragmentProjectBinding
import com.example.practice.viewmodel.MainViewModel
import com.example.practice.widge.MySlidingTabLayout
import com.wljy.mvvmlibrary.annotation.Event
import java.security.Key
import java.util.*
import java.util.ArrayList

/**
 * A simple [Fragment] subclass.
 * create an instance of this fragment.
 */
class ProjectFragment : BaseFragment<FragmentProjectBinding>() {


    var tableLayout: MySlidingTabLayout? = null
    var viewPager: ViewPager? = null
    var mainViewModel:MainViewModel?=null
    var projectListBeans:MutableList<ProjectListBean>?=null
    var projectPageBeans:MutableList<ProjectPageBean>?=null

    override fun initViewModel() {
        mainViewModel=registerViewModel(MainViewModel::class.java)
    }

    override fun initView(state: Bundle?) {
        super.initView(state)
        tableLayout=binding.tableLayout
        viewPager=binding.viewPager
    }

    override fun getRemoteData() {
        mainViewModel?.getProject()
    }

    @Event(key=[Constants.GET_PROJECT_LIST,Constants.REQUEST_ERROR])
    fun event(key: String,value :Any){
        if(key==Constants.GET_PROJECT_LIST){
            projectListBeans=value as MutableList<ProjectListBean>
            changePageFragment()
            relateVpAndTab()
        }else if(key==Constants.REQUEST_ERROR){

        }
    }

    private fun relateVpAndTab() {
        //      https://www.jianshu.com/p/8fdd6f2719b0
        val pageAdapter =ProjectPageAdapter(childFragmentManager,projectPageBeans!!)
        var titles = arrayOfNulls<String>(projectPageBeans?.size?:0)
        for(i in titles.indices){
            titles[i]= projectListBeans?.get(i)?.name
        }
        viewPager?.adapter=pageAdapter
        tableLayout?.setViewPager(viewPager,titles)
    }

    private fun changePageFragment():MutableList<ProjectPageBean> {
        projectPageBeans= ArrayList<ProjectPageBean>()
        projectListBeans?.forEach {
            val projectPageBean :ProjectPageBean= ProjectPageBean()
            projectPageBean.id=it.id
            projectPageBean.title=it.name
            projectPageBean.fragment=ProjectPageFragment.newInstance(it.id)
            projectPageBeans?.add(projectPageBean)
        }
        return projectPageBeans as ArrayList<ProjectPageBean>
    }

}